### PR TITLE
Added example script for InternVL

### PR DESCRIPTION
**Pull request by quic-dhirajku**

### CODE REVIEW

File: examples/internvl_inference.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+
+def run_intern_on_aic(model_name, prompt, image_url, messages, roles):
+    ## STEP 1 -- LOAD THE MODEL
+
+    model = QEFFAutoModelForCausalLM.from_pretrained(model_name, kv_offload=False, trust_remote_code=True)
+
+    ## STEP 2 -- EXPORT & COMPILE THE MODEL
+
+    model.compile(num_cores=14, num_devices=4, mxfp6_matmul=True)
```
Comment (quic-amitraj) [inline examples/internvl_inference.py:166]: Please update the num_cores to 16. We should always by default set the num_cores to 16 only.

File: examples/internvl_inference.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+    ## STEP 3 -- SETUP THE PROCESSOR
+
+    # InternVL doesn't have an AutoProcessor yet, so we will use our own processor class "InternProcessor"
+    qeff_pt_model = model.model
```
Comment (quic-amitraj) [inline examples/internvl_inference.py:171]: Is there any need of this line?

File: examples/internvl_inference.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+    # Resizing the image to have num_crops=13
+    # We can fix a different size and the compile the model with appropriate `num_crops`
+    image = image.resize((1000, 747))
+
+    # preprocess the resized image
+    pixel_values = internProcessor.load_image(image, max_num=12)
+    question = "<image>\n" + prompt
+    query = internProcessor(pixel_values, question, messages, roles)
+    inputs = tokenizer(query, return_tensors="pt", padding="max_length", max_length=3840, padding_side="right")
```
Comment (quic-amitraj) [inline examples/internvl_inference.py:188]: How you are setting max_length to `3840` if you are not setting the `prefil_seq_length` at the time of compilation?
There is two thing you can do-
1. Pass prefill seq as a arg to compile, so you will know the `max_length` value what to use at the time of tokenizer.
2. Else don't pass anything here, by default padding will be done till the `3840` as its the default value at the time of compilation.

File: examples/internvl_inference.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+    img = requests.get(image_url, stream=True)
+    image = Image.open(BytesIO(img.content)).convert("RGB")
+
+    # Resizing the image to have num_crops=13
+    # We can fix a different size and the compile the model with appropriate `num_crops`
+    image = image.resize((1000, 747))
+
+    # preprocess the resized image
+    pixel_values = internProcessor.load_image(image, max_num=12)
```
Comment (quic-amitraj) [inline examples/internvl_inference.py:185]: Change `max_num` to `num_patches`.

**Review by quic-amitraj — CHANGES_REQUESTED** [review]

File: examples/internvl_inference.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            if type(message) is tuple:
+                message, _, _ = message
+            ret += role + message + sep
+        else:
+            ret += role
+    return ret
+
+
+# Processor class for InternVL models
+class InternProcessor:
```
Comment (ochougul) [inline examples/internvl_inference.py:44]: write a docstring explaining this is taken from model card on hf and summarize the changes done to original code.

File: examples/internvl_inference.py (hunk 1 above)
Comment (quic-dhirajku) [reply examples/internvl_inference.py]: Updated, thanks.

File: examples/internvl_inference.py (hunk 2 above)
Comment (quic-dhirajku) [reply examples/internvl_inference.py]: Removed accrodingly.

File: examples/internvl_inference.py (hunk 3 above)
Comment (quic-dhirajku) [reply examples/internvl_inference.py]: Added `prefill_seq_len` as an added arguement now, thanks.

File: README.md [hunk 6]
```diff
@@ -6,18 +6,19 @@
 ---
 
 *Latest news* :fire: <br>
+- [02/2025] [VLMs support](https://github.com/quic/efficient-transformers/pull/267) added for the models [InternVL-1B](https://huggingface.co/OpenGVLab/InternVL2_5-1B), [Llava](https://huggingface.co/llava-hf/llava-1.5-7b-hf) (Single QPC) and [Mllama](https://huggingface.co/meta-llama/Llama-3.2-11B-Vision-Instruct) (Single + Dual QPCs)
```
Comment (ochougul) [inline README.md:9]: Single/dual qpc is internal terminology. We expose this via `kv_offload` variable.
Remove reference to single/dual qpc term

**Review by ochougul — CHANGES_REQUESTED** [review]

File: examples/internvl_inference.py (hunk 5 above)
Comment (quic-dhirajku) [reply examples/internvl_inference.py:44]: Done, thanks

File: README.md (hunk 6 above)
Comment (quic-dhirajku) [reply README.md:9]: Changes incorporated

File: examples/intern_example/internvl_inference.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True, use_fast=False)
+    internProcessor = InternProcessor(model.model, tokenizer)
+
+    ## STEP 4 -- PREPROCESS THE INPUTS
+
+    img = requests.get(image_url, stream=True)
+    image = Image.open(BytesIO(img.content)).convert("RGB")
+
+    # Resizing the image to have num_crops=13
+    # We can fix a different size and the compile the model with appropriate `num_crops`
```
Comment (ochougul) [inline examples/intern_example/internvl_inference.py:187]: `num_patches`

File: examples/vlm_inference.py
Comment (ochougul) [inline examples/vlm_inference.py:1]: rename to `text_to_image_inference.py`

File: examples/intern_example/internvl_inference.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    return ret
+
+
+# Processor class for InternVL models
+class InternProcessor:
+    """
+    InternVL model only has an AutoTokenizer so this class performs the processing tasks similar to an AutoProcessor.
+    The methods used here are borrowed from the original InternVL modelling files.
+    "https://huggingface.co/OpenGVLab/InternVL2_5-1B/"
+
```
Comment (ochougul) [inline examples/intern_example/internvl_inference.py:49]: remove line

File: examples/intern_example/internvl_inference.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        assert pixel_values is None or len(pixel_values) == sum(num_patches_list)
+        img_context_token_id = self.tokenizer.convert_tokens_to_ids(IMG_CONTEXT_TOKEN)
+        self.model.img_context_token_id = img_context_token_id
+
+        messages.append([roles[0], question])
+        messages.append([roles[1], None])
+        query = get_prompt(messages)
+        if verbose and pixel_values is not None:
+            image_bs = pixel_values.shape[0]
+            print(f"dynamic ViT batch size: {image_bs}")
```
Comment (ochougul) [inline examples/intern_example/internvl_inference.py:159]: why? use logger.info if needed.
Avoid using print statements

File: examples/intern_example/internvl_inference.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+    # Resizing the image to have num_crops=13
+    # We can fix a different size and the compile the model with appropriate `num_crops`
+    image = image.resize((1000, 747))
+
+    # preprocess the resized image
+    pixel_values = internProcessor.load_image(image, max_num=12)
+    question = "<image>\n" + prompt
+    query = internProcessor(pixel_values, question, messages, roles)
+    # inputs = tokenizer(query, return_tensors="pt", padding="max_length", padding_side="right")
```
Comment (ochougul) [inline examples/intern_example/internvl_inference.py:194]: commented code?

File: examples/intern_example/internvl_inference.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    model_name = "OpenGVLab/InternVL2_5-1B"
+
+    # Chat Template information for prompt preprocessing
+    messages: List[List[str]] = []
+    roles = ("<|im_start|>user\n", "<|im_start|>assistant\n")
+
+    # Inputs for the model
+    prompt = "Please describe the image in detail."
+    image_url = "https://image.slidesharecdn.com/azureintroduction-191206101932/75/Introduction-to-Microsoft-Azure-Cloud-1-2048.jpg"
+    prefill_seq_len = 3840
```
Comment (ochougul) [inline examples/intern_example/internvl_inference.py:216]: Add the explanation of why this is put as 3840

File: examples/vlm_inference.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    prefill_seq_len=32,
+    ctx_len=512,
+    img_size=448,
+    num_cores=16,
+    num_devices=1,
+):
+    ## STEP - 1 Load the Processor and Model
+
+    processor = AutoProcessor.from_pretrained(model_name, token=token)
+    # `kv_offload` is used to decide if we wish to run Single QPC or 2 QPC setup
```
Comment (ochougul) [inline examples/vlm_inference.py:33]: single/ dual qpc is internal term. can you explain properly here

File: examples/vlm_inference.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        token=HF_TOKEN,
+        query=query,
+        kv_offload=kv_offload,
+        image_url=image_url,
+        prefill_seq_len=prefill_seq_len,
+        ctx_len=ctx_len,
+        img_size=img_size,
+        num_cores=num_cores,
+        num_devices=num_devices,
+    )
```
Comment (ochougul) [inline examples/vlm_inference.py:103]: What's the expected output for this model? can you put that here?

**Review by ochougul — CHANGES_REQUESTED** [review]

File: examples/intern_example/internvl_inference.py (hunk 8 above)
Comment (ochougul) [reply examples/intern_example/internvl_inference.py:49]: empty line

Comment (ochougul) [reply examples/intern_example/internvl_inference.py:49]: # 49

File: examples/intern_example/internvl_inference.py [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+    tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True, use_fast=False)
+    internProcessor = InternProcessor(model.model, tokenizer)
+
+    ## STEP 4 -- PREPROCESS THE INPUTS
+
+    img = requests.get(image_url, stream=True)
+    image = Image.open(BytesIO(img.content)).convert("RGB")
+
+    # Resizing the image to have num_patches=13
+    # We can fix a different size and the compile the model with appropriate `num_patches`
```
Comment (quic-rishinr) [inline examples/intern_example/internvl_inference.py:203]: Please add some more details on how to provide different image sizes and select the num_patches.

File: examples/intern_example/internvl_inference.py [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+    messages,
+    roles,
+    kv_offload=False,
+    prefill_seq_len=3840,
+    num_devices=1,
+    num_cores=16,
+):
+    ## STEP 1 -- LOAD THE MODEL
+
+    model = QEFFAutoModelForCausalLM.from_pretrained(model_name, kv_offload=kv_offload, trust_remote_code=True)
```
Comment (quic-rishinr) [inline examples/intern_example/internvl_inference.py:180]: Add a comment why QEFFAutoModelForCausalLM use for a VLM.

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

**Review by quic-rishinr — APPROVED** [review]

File: examples/image_text_to_text_inference.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+    query = "Describe this image."
+    image_url = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/0052a70beed5bf71b92610a43a52df6d286cd5f3/diffusers/rabbit.jpg"
+
+    # Compilation parameters for the model
+    kv_offload = False
+    prefill_seq_len = 32
+    ctx_len = 512
+    img_size = 560
+    num_cores = 16
+    num_devices = 1
```
Comment (ochougul) [inline examples/image_text_to_text_inference.py:94]: can you please define `generation_len` here and pass it to `run_model` ?
everything else LGTM.
Will merge after this change

Comment (quic-dhirajku) [reply examples/image_text_to_text_inference.py:94]: sure, will do, thanks.

**Review by ochougul — APPROVED** [review]
